Procesamiento del Lenguaje Natural
===
Lab 1 - NLTK y Contar Palabras Corpora
===
**Alumno:** Gonzalo Bertinat

**Legajo:** 160.615-3

**Curso:** K3572

**Probando sus conocimientos**
===

Veamos la primer pregunta del cuestionario...

**1. ¿Cuál es el número de tokens en Moby Dick?**

Antes de ponernos a calcular dicho número, vamos a importar lo que consideremos necesario...

Primero, hay que importar el texto de Moby Dick! 
Para eso, como se nos explicó a lo largo del instructivo del Lab debemos realizar lo siguiente:

In [1]:
# Importamos los textos existentes desde NLTK
from nltk.book import *

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


In [2]:
# Como ya se aclaró en el instructivo, Moby Dick se corresponde con el text1.
# Guardamos en una variable los tokens...
moby_dick_tokens = text1.tokens

Ahora uno estaría tentado a poder contestar la pregunta, ¿no?

Acaso no tenemos los tokens? Tranquilamente podríamos hacer:

In [3]:
len(moby_dick_tokens)

260819

Y listo, ¿verdad?

Bueno, la realidad es que **esto está mal.** 

El enunciado nos aclara antes de preguntarnos cualquier cosa que:
    
**Remueva las puntuaciones y convierta a minúsculas (lower case) antes de hacer las cuentas.**

Mejor hagamos caso...

In [4]:
# Importamos re para poder utilizar expresiones regulares...
from nltk import re
# Mediante re.search("\w",word) matcheamos solo aquellos tokens que son alfanuméricos.
# Con el for iteramos sobre todos los tokens de moby_dick_tokens y los convertimos a minúsculas con lower()
moby_dick_tokens_nopunct = [word.lower() for word in moby_dick_tokens if re.search("\w",word)]                            

Genial! Ahora sí tenemos los tokens como los pide el enunciado.

Estamos en condiciones de pedir la cantidad:

In [5]:
cantidad_tokens_MD = len(moby_dick_tokens_nopunct)
print("La cantidad de tokens de Moby Dick es:", cantidad_tokens_MD)

La cantidad de tokens de Moby Dick es: 218621


Excelente!

Veamos la siguiente:

**2. ¿Cuál es el número de types en Moby Dick?**

Bueno, para esto primero recordemos que los **types** no eran nada menos que los tokens pero quitando sus repeticiones, es decir, **el número de palabras únicas**.

Afortunadamente, para quitar los tokens repetidos que tenemos guardados en la variable **moby_dick_tokens_nopunct** vamos a utilizar el set de Python, que justamente, crea una colección de elementos SIN REPETIR. 

In [6]:
moby_dick_types = set(moby_dick_tokens_nopunct)

Ahora podemos contestar la pregunta 2:

In [7]:
cantidad_types_MD = len(moby_dick_types)
print("La cantidad de Types en Moby Dick es:",cantidad_types_MD)

La cantidad de Types en Moby Dick es: 17140


**3. Moby Dick type-token ratio**

Ahora se nos pide este ratio... Pero, ¿qué es?

Simplemente es dividir la cantidad de types por la cantidad de tokens. 

Lo que hace es darnos una idea de la diversidad léxica de un texto, es decir que tan diverso en palabras es en relación a la cantidad total de palabras que lo componen.

Se nos pide el TTR (Type-Token Ratio) de Moby Dick, y afortunadamente de lo que ya hicimos antes tenemos guardados en variables los dos valores que necesitamos: la cantidad de tokens y la cantidad de types.

Entonces:

In [8]:
moby_dick_TTR = cantidad_types_MD / cantidad_tokens_MD
print("El Type-Token ratio de Moby Dick es:",moby_dick_TTR)

El Type-Token ratio de Moby Dick es: 0.07840051962071347


**4. WSJ type-token ratio**

Para esta tarea repetimos el mismo código que realizamos con Moby Dick, pero ahora con el texto de Wall Street Journal (WSJ).

La única diferencia es que como el texto es otro, debemos pedir los tokens en este caso al texto 7:

In [9]:
# Pedimos los tokens
WSJ_tokens = text7.tokens

# Quitamos tokens de puntuación y convertimos a minúsculas 
WSJ_tokens_no_punct = [word.lower() for word in WSJ_tokens if re.search("\w",word)]

# Cantidad de tokens final
cantidad_tokens_WSJ = len(WSJ_tokens_no_punct)

# Verificamos esa cantidad:
print("La cantidad de tokens en Wall Street Journal es",cantidad_tokens_WSJ)

# Ahora armamos un set para eliminar repetidos y pedimos la cantidad para obtener la cantidad de types
cantidad_types_WSJ = len(set(WSJ_tokens_no_punct))
print("La cantidad de types en Wall Street Journal es:",cantidad_types_WSJ)

# Calculamos el TTR de WSJ 
WSJ_TTR = cantidad_types_WSJ/cantidad_tokens_WSJ
print("El Type-Token Ratio de Wall Street Journal es:",WSJ_TTR)

La cantidad de tokens en Wall Street Journal es 87608
La cantidad de types en Wall Street Journal es: 11367
El Type-Token Ratio de Wall Street Journal es: 0.129748424801388


**5. ¿Cuál de los dos tiene más diversidad léxica?**

Wall Street Journal con 0.129748424801388

Pero para hacerlo más divertido y con código Python, mejor se lo preguntamos a él:

In [10]:
if moby_dick_TTR > WSJ_TTR:
    print("Moby Dick tiene mayor diversidad léxica! Su Type-Token Ratio fue:",moby_dick_TTR)
elif WSJ_TTR > moby_dick_TTR:
    print("Wall Street Journal tiene mayor diversidad léxica! Su Type-Token Ratio fue:",WSJ_TTR)
else:
    print("Moby Dick y Wall Street Journal son igualmente diversos léxicamente!")

Wall Street Journal tiene mayor diversidad léxica! Su Type-Token Ratio fue: 0.129748424801388


**6. ¿Puede pensar una razón por por la cual ese corpus es más diverso que el otro?**

Si tengo que responder en frío esta pregunta le haría caso a mi profesor respondiendo un **"Depende"**. Pero hay que razonar un poco según lo que se hizo hasta ahora en el lab.

Haciendo un análisis rápido, puedo llegar a la siguiente conclusión:

Wall Street Journal tiene menos cantidad de palabras, por lo que la probabilidad de que repita palabras
es menor, y si se repiten menos las palabras, el ratio empezará a crecer, tendremos más types.
Creo que es natural que cuanto más escribamos, más repetimos palabras de uso común, como por ejemplo las 
palabras "el", "la", "si", "a", etcétera. Por eso considero que Moby Dick tiene un ratio menor, pensar solamente
en cuántas veces aparece la palabra "Moby", debe haber bastantes si de un personaje se trata.
 
Espero que sea correcto, aunque insisto que depende mucho del texto que se esté analizando, un autor tiene tendencia a escribir de una misma forma, por lo que comenzará a repetir léxico y se hará menos diverso, eso también podría explicar que Moby Dick tenga menor ratio. Por otro lado Wall Street Journal contempla noticias, las cuales a priori deberían ser de muchos tópicos, incorporando muchas palabras distintas y dandole más ratio que a Moby Dick.


**7. ¿Cuál es el “Maximum Likelihood Estimate (MLE)” de la palabra “whale” (ballena) en Moby Dick?**

Para responder esto, recordemos que el Maximum Likelihood Estimate de una palabra se podía calcular en Unigram como la cantidad de apariciones de esa palabra en el corpus dividida por la cantidad de palabras del corpus. 

En unigram, el MLE es equivalente a la probabilidad de que una palabra aparezca en el corpus.

$$ MLE(wi)=\frac{count(wi)}{N} $$

Dejemos que Python se encargue...

In [11]:
# Con el método count() calculamos la cantidad de apariciones en el corpus
cantidad_apariciones_whale_MD = moby_dick_tokens_nopunct.count("whale")
# Aplicamos la fórmula del MLE
MLE_whale_moby_dick = cantidad_apariciones_whale_MD / cantidad_tokens_MD

print("El Maximum Likelihood Estimate de la palabra 'whale' en Moby Dick es:",MLE_whale_moby_dick)
print("Pmoby dick(“whale”) =",MLE_whale_moby_dick)

El Maximum Likelihood Estimate de la palabra 'whale' en Moby Dick es: 0.005607878474620462
Pmoby dick(“whale”) = 0.005607878474620462


**8. ¿Cuál es el MLE de “whale” en el corpus de WSJ?**

Planteamos exactamente lo mismo que para Moby Dick:

In [12]:
cantidad_apariciones_whale_WSJ = WSJ_tokens_no_punct.count("whale")
MLE_whale_WSJ = cantidad_apariciones_whale_WSJ / cantidad_tokens_WSJ
print("El Maximum Likelihood Estimate de la palabra 'whale' en Wall Street Journal es:",MLE_whale_WSJ)
print("Pmoby dick(“whale”) =",MLE_whale_WSJ)

El Maximum Likelihood Estimate de la palabra 'whale' en Wall Street Journal es: 0.0
Pmoby dick(“whale”) = 0.0


Finalmente, repasando todas las respuestas:

1. La cantidad de tokens de Moby Dick es: 218621
2. La cantidad de Types en Moby Dick es: 17140
3. El Type-Token ratio de Moby Dick es: 0.07840051962071347
4. El Type-Token Ratio de Wall Street Journal es: 0.129748424801388
5. Wall Street Journal tiene mayor diversidad léxica que Moby Dick
6. Contestada ampliamente más arriba
7. El Maximum Likelihood Estimate de la palabra 'whale' en Moby Dick es: 0.005607878474620462
8. El Maximum Likelihood Estimate de la palabra 'whale' en Wall Street Journal es: 0.0

In [13]:
print("Esto fue todo amigos!")

Esto fue todo amigos!


FIN DEL NOTEBOOK
==